src/dataframe/dataframe.jl

```diff
@@ -1094,9 +1114,56 @@ function permutecols!(df::DataFrame, p::AbstractVector)
         throw(ArgumentError("$p is not a valid column permutation for this DataFrame"))
     end
     permute!(columns(df), p)
+    permute!(metadata(df), p)
     setfield!(df, :colindex, Index(names(df)[p]))
 end
 
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, getindex.(index(df).lookup, p))
 end
+
+
+##############################################################################
+##
+## Set and Get MetaData 
+##
+##############################################################################
+
+"""
+    addlabel!(df::DataFrame, var::Symbol, label::String)
```

Comment:
Is this just for now, or do you envision this in the future? Because an idiomatic way to set labels, particularly with chaining in mind, first argument is a dataframe, returns a dataframe, seems important.

Reply:
Another thing with this approach is that `metadata` doesn't know anything about the symbols and the `colindex` of the dataframe. So any function would have to include the `dataframe` in it, right?

Reply:
> Is this just for now, or do you envision this in the future? Because an idiomatic way to set labels, particularly with chaining in mind, first argument is a dataframe, returns a dataframe, seems important.

I don't know yet, that's why I'd rather start with the strictly minimal API.

Reply:
I also changed `getmeta` and `showmeta` to be more generic. `showmeta` now returns a dataframe, but its probably a bad idea cause dataframes aren't that readable for long strings. But it was easy to write and not a horrible idea.

Reply:
As I said, let's start with a minimal implementation. We can always add convenience method later if that's useful.

I'd also rather rename `getmeta` to `metadata` and `addmeta!` to `setmetadata!` or `metadata!`.

Reply:
I understand what you are saying more about `setindex`. Might be nice to write 

`metadata(df, :x1, :label) = "A variable label"` work.